File: metal_mps/matmul_torch.cc
#include <chrono>
#include <iostream>

#include <torch/mps.h>
#include <torch/torch.h>

// be consistent in all tests.
constexpr size_t SIZE = 4096L * 2L;

const auto deviceToUse = torch::kMPS;
// const auto deviceToUse = torch::kCPU;

int
main()
{
        auto opt =
            torch::TensorOptions().dtype(torch::kFloat32).device(deviceToUse);
        torch::Tensor a = torch::randn({SIZE, SIZE}, opt);
        torch::Tensor b = torch::randn({SIZE, SIZE}, opt);
        torch::Tensor c = torch::randn({SIZE, SIZE}, opt);

        //
        // warm up
        // compute C = matmul(A, B)
        //
        torch::mps::synchronize();

        auto count = 10;
        for (int i = 0; i < count; i++) {
                at::matmul_out(c, a, b);
        }

        torch::mps::synchronize();
        //
        // real game
        // compute C = matmul(A, B)
        //
        auto start = std::chrono::system_clock::now();

        for (int i = 0; i < count; i++) {
                at::matmul_out(c, a, b);
        }

        torch::mps::synchronize();
        auto end = std::chrono::system_clock::now();

        std::chrono::duration<double, std::milli> fp_ms = end - start;

        std::cout << "libtorch device " << deviceToUse << "\n";
        std::cout << "c.matmul takes " << fp_ms.count() / count / 1000 << "\n";
        return 0;
}

File: metal_mps/matmul_torch.py
import torch
import time

print(f"torch mps backend: {torch.backends.mps.is_available()}")

# choose a device
device = "cpu"
#device = "mps"

# shapes must be consistent in all tests.
a = torch.randn([4096 * 2, 4096 * 2], device=device)
b = torch.randn([4096 * 2, 4096 * 2], device=device)

def sync():
    if device == "mps":
        torch._C._mps_synchronize()

sync()

# warm up
for i in range(10):
    c = torch.matmul(a, b)

# real game
count = 10
sync()
start = time.time()
for i in range(count):
    c = torch.matmul(a, b)
sync()
end = time.time()

print(f"c.device is {c.device}")
print(f"c.dtype is {c.dtype}")
print(f"c is {c}")
print(f"c.matmul takes {(end-start) / count }")
